cri: fix issues caused by converting the image reference

Iceber opened the pull request:
fix: https://github.com/containerd/containerd/issues/8848

As https://github.com/containerd/containerd/issues/8848#issuecomment-1649211628, the cri plugin's conversion of containerd image references causes a lot of issues.

After thinking about multiple implementations, I want the cri plugin to ignore images whose references are converted, and these images do not set the `io.cri-containerd.image=managed` label.

Iceber commented on pkg/cri/store/image/image.go:
```diff
@@ -212,7 +212,7 @@ func (s *store) add(img Image) error {
 	}
 	// Or else, merge and sort the references.
-	i.References = docker.Sort(util.MergeStringSlices(i.References, img.References))
+	i.References = docker.StrictSort(util.MergeStringSlices(i.References, img.References))
```
Now the image store doesn't modify the image ref, which fixes an inconsistency between the `refCache` and `image.References`.

Although inconsistent images are already ignored in the `cri plugin service`, I think it's still necessary to fix the bug in the image store.

Iceber commented:
/test pull-containerd-node-e2e

upodroid commented:
hi, the failure in `pull-containerd-node-e2e` is being fixed in #8862

Iceber commented:
@fuweid @thaJeztah PTAL, Thanks

fuweid commented on pkg/cri/sbserver/images/image_pull.go:
```diff
@@ -328,6 +328,11 @@ func (c *CRIImageService) getLabels(ctx context.Context, name string) map[string
 // in containerd. If the reference is not managed by the cri plugin, the function also
 // generates necessary metadata for the image and make it managed.
 func (c *CRIImageService) UpdateImage(ctx context.Context, r string) error {
```
would you mind to test with starting pod with `busybox` image? The image is without domain name and `library` namespace. I remembered that kubelet won't add the default domain name. So I want to confirm that it doesn't break anything. thanks

Iceber replied:
it's ok, kubelet does not add the default domain prefix, but the `cri plugin` converts `busybox:latest` to `docker.io/library/busybox:latest`

After the pod starts, use crictl and ctr to view the image
```bash
$ ctr -n k8s.io images | grep busybox                                               root@caiwei-master:~# ctr -n k8s.io images ls | grep busybox
docker.io/library/busybox:latest                                                                                                        application/vnd.docker.distribution.manifest.list.v2+json sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79 2.1 MiB   linux/386,linux/amd64,linux/arm/v5,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x io.cri-containerd.image=managed
docker.io/library/busybox@sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79                                       application/vnd.docker.distribution.manifest.list.v2+json sha256:3fbc632167424a6d997e74f52b878d7cc478225cffac6bc977eedfe51c7f4e79 2.1 MiB   linux/386,linux/amd64,linux/arm/v5,linux/arm/v6,linux/arm/v7,linux/arm64/v8,linux/mips64le,linux/ppc64le,linux/riscv64,linux/s390x io.cri-containerd.image=managed
$ crictl images | grep busybox
docker.io/library/busybox                                         latest              a416a98b71e22       2.22MB
```

Iceber commented on reference/docker/sort.go:
```diff
@@ -56,6 +56,43 @@ func Sort(references []string) []string {
+// 5. Digest
+// 6. Parse error and inconsistent ref
+// Unlike Sort, it does not modify the items of `references`
+func StrictSort(references []string) []string {
```
The best way to do this, would be to remove `Sort` and rename `StrictSort` to `Sort`, but this would be an incompatible change for the `Sort` function, so I kept `Sort` and marked it as deprecated

akhilerm commented:
> I want the cri plugin to ignore images whose references are converted, and these images do not set the io.cri-containerd.image=managed label.

Does this mean , if I tag an image without the domain, i.e
```shell
ctr -n k8s.io images tag docker.io/library/busybox:1.36 busybox:fixed
```
the image `busybox:fixed` will not have the `io.cri-containerd.image=managed` label

Iceber commented:
> > I want the cri plugin to ignore images whose references are converted, and these images do not set the io.cri-containerd.image=managed label.
> 
> Does this mean , if I tag an image without the domain, i.e
> 
> ```shell
> ctr -n k8s.io images tag docker.io/library/busybox:1.36 busybox:fixed
> ```
> 
> the image `busybox:fixed` will not have the `io.cri-containerd.image=managed` label

Yes, since the cri plugin can't manage such images, they won't be labeled

fuweid commented:
Not sure that it breaks air gapped usage. Changes look good. But I think it needs more reviewers on this.

Iceber commented on pkg/cri/server/image_pull.go:
@akhilerm Thanks, I overlooked the case where the image already has this label.

This label is managed by cri, not a user-defined label, so I think we need to remove this label for the undomain image.

I updated pr, ptla
